File: w3/dz_cpp/bcw4/bcw4/Basic/Vector/Vector.cpp
#include <iostream>
#include <cmath>
#include "Vector.h"

Vector::Vector(double x, double y) {
    this->x = x;
    this->y = y;
}

Vector::~Vector() {}

double Vector::getX() const {
    return this->x;
}

double Vector::getY() const {
    return this->y;
}

void Vector::setX(double x) {
    this->x = x;
}

void Vector::setY(double y) {
    this->y = y;
}

double Vector::len() const {
    return hypot(this->x, this->y);
}

bool Vector::operator==(const Vector& other) const {
    return this->x == other.x && this->y == other.y;
}

bool Vector::operator!=(const Vector& other) const {
    return !(*this == other);
}

void Vector::operator+=(const Vector& other) {
    this->x += other.x;
    this->y += other.y;
}

void Vector::operator-=(const Vector& other) {
    this->x -= other.x;
    this->y -= other.y;
}

Vector Vector::operator+(const Vector& other) const {
    Vector result = *this;

    result += other;
    return result;
}

Vector Vector::operator-(const Vector& other) const {
    Vector result = *this;

    result -= other;
    return result;
}

std::ostream& operator<<(std::ostream& out, const Vector& vector) {
    return out << "(" << vector.getX() << ", " << vector.getY() << ")";
}

File: w3/dz_java/src/Basic/Vector/Vector.java
package Basic.Vector;

import static java.lang.Math.hypot;

public class Vector {
    private double x;
    private double y;

    public Vector() {
        this.x = 0;
        this.y = 0;
    }

    public Vector(double x, double y) {
        this.x = x;
        this.y = y;
    }

    public double getX() {
        return x;
    }

    public double getY() {
        return y;
    }

    public void setX(double x) {
        this.x = x;
    }

    public void setY(double y) {
        this.y = y;
    }

    public double len() {
        return hypot(this.x, this.y);
    }

    public boolean equal(Vector other) {
        return this.x == other.x && this.y == other.y;
    }

    public boolean notEqual(Vector other) {
        return !this.equal(other);
    }

    public void addition(Vector other) {
        this.x += other.x;
        this.y += other.y;
    }

    public void substraction(Vector other) {
        this.x -= other.x;
        this.y -= other.y;
    }

    public Vector sum(Vector other) {
        Vector result = this;

        result.addition(other);
        return result;
    }

    public Vector diff(Vector other) {
        Vector result = this;

        result.substraction(other);
        return result;
    }

    public String toString() {
        return String.format( "(" + this.getX() + ", " + this.getY() + ")" );
    }
}
